votingsystem/src/voting.rs

```diff
@@ -0,0 +1,419 @@
+        let vote_paper_id = vote.corresponding_paper_id;
+        let vote_paper: VotePaper =
+            serde_cbor::from_slice(&self.storage().get(vote_paper_id.0.as_ref()).unwrap()).unwrap();
+        Ok(vote_paper.number_of_shares)
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct TxCreateVotePaper {
+    general_meeting_id: GeneralMeetingId,
+    agenda_number: u32,
+    number_of_shares: u32,
+    voter_name: String,
+    voter_public_key: Public,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct TxVote {
+    vote_paper_id: H256,
+    choice: String,
+    voter_signature: Signature,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub enum TxAction {
+    CreateVotePaper(TxCreateVotePaper),
+    Vote(TxVote),
+}
+
+impl Action for TxAction {}
+pub type OwnTransaction = crate::common::SignedTransaction<TxAction>;
+
+impl Context {
+    fn excute_tx(&mut self, transaction: &Transaction) -> Result<H256, ExecuteError> {
+        if transaction.tx_type() != "Vote" {
+            return Err(ExecuteError::InvalidMetadata)
+        }
+        let tx: OwnTransaction =
+            serde_cbor::from_slice(&transaction.body()).map_err(|_| ExecuteError::InvalidFormat)?;
+        tx.verify().map_err(|_| ExecuteError::InvalidSign)?;
+
+        match tx.tx.action {
+            TxAction::CreateVotePaper(TxCreateVotePaper {
+                general_meeting_id,
+                agenda_number,
+                number_of_shares,
+                voter_name: voter_name,
+                voter_public_key,
+            }) => {
+                let meeting: GeneralMeeting = self.meeting_mut().get_meeting(&general_meeting_id).unwrap();
+                if meeting.id != general_meeting_id {
+                    return Err(ExecuteError::InvalidMeeting)
+                }
+                if meeting.tallying_time.time < self.block_header.as_ref().unwrap().timestamp() {
+                    return Err(ExecuteError::InvalidMeetingTime)
+                }
+                if meeting.number_of_agendas < agenda_number {
+                    return Err(ExecuteError::InvalidNumberofAgendas)
+                }
+                if meeting.voting_end_time.time > self.block_header.as_ref().unwrap().timestamp() {
+                    return Err(ExecuteError::InvalidVotingTime)
+                }
+                let vote_paper = VotePaper::new(
+                    general_meeting_id.clone(),
+                    agenda_number,
+                    voter_name,
+                    number_of_shares,
+                    voter_public_key,
+                );
+                self.storage_mut().set(vote_paper.vote_paper_id.0.as_ref(), serde_cbor::to_vec(&vote_paper).unwrap());
+                Ok(vote_paper.vote_paper_id)
+            }
+
+            TxAction::Vote(TxVote {
+                vote_paper_id,
+                choice,
+                voter_signature,
```

Comment:
Please check the voter_signature is generated from the voter private key.
Since we have voter public key, we can verify it.

Reply:
@MSNTCS The verification happens when `execute_tx` is called.